uncertainties/test_uncertainties.py

```diff
@@ -2338,3 +2338,40 @@ def test_correlated_values_correlation_mat():
         assert arrays_close(
             numpy.array(cov_mat),
             numpy.array(uncert_core.covariance_matrix([x2, y2, z2])))
+
+    def test_hash():
+        '''
+        Tests the invariance that if x==y, then hash(x)==hash(y)
+        '''
+
+        a = ufloat(1.23, 2.34)
+        b = ufloat(1.23, 2.34)
+
+        # nominal values and std_dev terms are equal, but...
+        assert a.n==b.n and a.s==b.s
+        # ...x and y are independent variables, therefore not equal as uncertain numbers
+        assert a != b
+        assert hash(a) != hash(b)
+
+        # order of calculation should be irrelevant
+        assert a + b == b + a
+        assert hash(a + b) == hash(b + a)
+
+        # the equation (2x+x)/3 is equal to the variable x, so...
+        assert ((2*a+a)/3)==a
+        # ...hash of the equation and the variable should be equal
+        assert hash((2*a+a)/3)==hash(a)
```

Comment:
This one is simple enough that it works out okay, but if the calculation got more complicated than `(2. + 1.) / 3 == 1.` the equality check could fail due to floating point rounding error. Luckily though the equality check would fail along with the hash, so it would not be a violation of the model.

Reply:
Yes, I think you are right.
I already wondered, why the equality check is implemented in such a strange way.
But I did not step into details. Probably there is a good reason for that.